lib/validation.rb

```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+module Validation
+  class ValidationFailed < StandardError
+    attr_reader :errors
+    def initialize(errors)
+      @errors = errors
+      super("Validation failed for some fields")
+    end
+  end
+
+  # Allow DNS compatible names
+  # - Length between 3 and 63
+  # - Only lowercase letters, numbers, and hyphens
+  # - Not start or end with a hyphen
+  # - Not only numbers
+  ALLOWED_NAME_PATTERN = '\A(?![0-9]+$)(?!-)[a-z0-9-]{3,63}(?<!-)\z'
```

Comment:
You should probably source this from something like https://stackoverflow.com/questions/106179/regular-expression-to-match-dns-hostname-or-ip-address. 

Another thing I was thinking about was to always change generated UUIDs to have a non-integral first digit.  This has an unfortunate conflict with ULIDs, which always generate a leading zero for quite a while. We could fix it with some byteswapping, but it's rather ugly.